pkg/controller/clusterdeployment/metrics.go

```diff
@@ -83,12 +62,67 @@ func incProvisionFailedTerminal(cd *hivev1.ClusterDeployment) {
 	if stoppedCondition != nil {
 		stoppedReason = stoppedCondition.Reason
 	}
-	metricProvisionFailedTerminal.WithLabelValues(poolNSName,
-		hivemetrics.GetClusterDeploymentType(cd),
-		stoppedReason).Inc()
+	metricProvisionFailedTerminal.FixedLabels["clusterpool_namespacedname"] = poolNSName
+	metricProvisionFailedTerminal.FixedLabels["failure_reason"] = stoppedReason
+	metricProvisionFailedTerminal.LogMetricWithDynamicLabels(cd, log, 1)
 }
 
-func init() {
+func registerMetrics(mapClusterTypeLabelToValue map[string]string) {
+	optionalClusterTypeLabels := hivemetrics.GetSortedKeys(mapClusterTypeLabelToValue)
+
+	metricCompletedInstallJobRestarts = hivemetrics.HistogramWithDynamicLabels{
+		HistogramVec: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name:    "hive_cluster_deployment_completed_install_restart",
+				Help:    "Distribution of the number of restarts for all completed cluster installations.",
+				Buckets: []float64{0, 2, 10, 20, 50},
+			},
+			optionalClusterTypeLabels),
+		FixedLabels:    map[string]string{},
+		OptionalLabels: mapClusterTypeLabelToValue,
+	}
+
+	metricClustersCreated = hivemetrics.CounterWithDynamicLabels{
+		CounterVec: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "hive_cluster_deployments_created_total",
+			Help: "Counter incremented every time we observe a new cluster.",
+		},
+			optionalClusterTypeLabels),
+		FixedLabels:    map[string]string{},
+		OptionalLabels: mapClusterTypeLabelToValue,
+	}
+	metricClustersInstalled = hivemetrics.CounterWithDynamicLabels{
+		CounterVec: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "hive_cluster_deployments_installed_total",
+			Help: "Counter incremented every time we observe a successful installation.",
+		},
+			optionalClusterTypeLabels),
+		FixedLabels:    map[string]string{},
+		OptionalLabels: mapClusterTypeLabelToValue,
+	}
+	metricClustersDeleted = hivemetrics.CounterWithDynamicLabels{
+		CounterVec: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "hive_cluster_deployments_deleted_total",
+			Help: "Counter incremented every time we observe a deleted cluster.",
+		},
+			optionalClusterTypeLabels),
+		FixedLabels:    map[string]string{},
+		OptionalLabels: mapClusterTypeLabelToValue,
+	}
+	metricProvisionFailedTerminal = hivemetrics.CounterWithDynamicLabels{
+		CounterVec: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "hive_cluster_deployments_provision_failed_terminal_total",
+			Help: "Counter incremented when a cluster provision has failed and won't be retried.",
+		},
+			append([]string{"clusterpool_namespacedname", "failure_reason"}, optionalClusterTypeLabels...)),
+		FixedLabels: map[string]string{
+			// These values must be filled before logging the metric
+			"clusterpool_namespacedname": constants.MetricLabelDefaultValue,
+			"failure_reason":             constants.MetricLabelDefaultValue,
```

Comment:
I wasn't sure if I should pass around an empty map here or do it this way. wdyt?

Reply:
Not sure what you're asking. I dig the idea of `FixedLabels` and `OptionalLabels`. Though possibly `FixedLabels` should just be a slice with the keys -- what's the value of having it be a map?

Reply:
[Later] Okay, now I see how you were using the map-ness. As noted, that won't work because globals.

I think we actually end up not needing the `FixedLabels` struct member at all.